homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,226 @@
+
+    return unload_ok
+
+
+class Coordinator(DataUpdateCoordinator[dict[int, Coil]]):
+    """Update coordinator for nibe heat pumps."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        heatpump: HeatPump,
+        connection: Connection,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, LOGGER, name="Nibe Heat Pump", update_interval=timedelta(seconds=60)
+        )
+
+        self.data = {}
+        self.connection = connection
+        self.heatpump = heatpump
+
+    @property
+    def coils(self) -> list[Coil]:
+        """Return the full coil database."""
+        return self.heatpump.get_coils()
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this coordinator."""
+        return self.config_entry.unique_id or self.config_entry.entry_id
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for the main device."""
+        return DeviceInfo(identifiers={(DOMAIN, self.unique_id)})
+
+    def get_coil_value(self, coil: Coil) -> int | str | float | None:
+        """Return a coil with data and check for validity."""
+        if coil := self.data.get(coil.address):
+            return coil.value
+        return None
+
+    def get_coil_float(self, coil: Coil) -> float | None:
+        """Return a coil with float and check for validity."""
+        if value := self.get_coil_value(coil):
+            return float(value)
+        return None
+
+    async def async_write_coil(
+        self, coil: Coil | None, value: int | float | str
+    ) -> None:
+        """Write coil and update state."""
+        if not coil:
+            raise HomeAssistantError("No coil available")
+
+        coil.value = value
+        coil = await self.connection.write_coil(coil)
+
+        if self.data:
+            self.data[coil.address] = coil
+            self.async_update_listeners()
+
+    async def _async_update_data(self) -> dict[int, Coil]:
+        @retry(
+            retry=retry_if_exception_type(CoilReadException), stop=stop_after_attempt(2)
+        )
+        async def read_coil(coil: Coil):
+            return await self.connection.read_coil(coil)
+
+        callbacks: dict[int, list[CALLBACK_TYPE]] = {}
+        for update_callback, context in list(self._listeners.values()):
+            assert isinstance(context, set)
+            for address in context:
+                callbacks.setdefault(address, []).append(update_callback)
+
+        result: dict[int, Coil] = {}
+
+        for address, callback_list in callbacks.items():
+            try:
+                coil = self.heatpump.get_coil_by_address(address)
+                self.data[coil.address] = result[coil.address] = await read_coil(coil)
+            except (CoilReadException, RetryError) as exception:
+                self.logger.warning("Failed to update: %s", exception)
```

Comment:
If we fail to update we normally raise `UpdateFailed`. Just logging here will spam the log and also not update `coordinator.last_update_success` correctly.

Reply:
We don't want to fail the whole update. Single address can time out from time to time without us having to kill the whole update. I could increase the number of retries i suppose to some really large number, and assume full failure after that.

Probably make sense. Will do that.

Reply:
https://github.com/home-assistant/core/pull/78931/commits/063a98fc2cea1f67804799a2feb34d4c5bf66366